torch/csrc/expand_utils.h

```diff
@@ -0,0 +1,155 @@
+#ifndef THP_EXPAND_UTILS_H
+#define THP_EXPAND_UTILS_H
+
+#include <sstream>
+#include <Python.h>
+
+template <typename ExpandType>
+ExpandType *newForExpand(LIBRARY_STATE_TYPE_NOARGS);
+
+template <typename TensorType>
+int expand(LIBRARY_STATE_TYPE TensorType *r, TensorType *tensor, THLongStorage *sizes, int raiseErrors);
+
+template <typename TensorType>
+int expand2(LIBRARY_STATE_TYPE TensorType *r1, TensorType *r2,
+            TensorType *e1, TensorType *e2, int raiseErrors);
+
+template <typename TensorType>
+int expand3(LIBRARY_STATE_TYPE TensorType *r1, TensorType *r2, TensorType *r3,
+            TensorType *e1, TensorType *e2, TensorType *e3, int raiseErrors);
+
+template <typename ExpandType, typename TensorType>
+void check_fallback(ExpandType *to_expand, TensorType *tensor,
+                    char *to_expand_name, char *tensor_name, bool fallback,
+                    ptrdiff_t to_expand_nElem, ptrdiff_t tensor_nElem, int to_expand_err) {
+  if (fallback && getBackCompatBroadcastWarn()) {
+    bool same_shape = THSize_isSameSizeAs(tensor->size, tensor->nDimension,
+        to_expand->size, to_expand->nDimension);
+    if (!same_shape && to_expand_err == 0 && (tensor_nElem == to_expand_nElem) && fallback) {
+      std::ostringstream warn;
+      warn << tensor_name << " and " << to_expand_name << " do not have the same shape, but are "
+           << "broadcastable, and have the same number of elements.  Changing behavior in a backwards incompatible "
+           << "manner to broadcasting rather than viewing as 1-dimensional.";
```

Comment:
Shouldn't we warn even if they are not broadcastable? I thought we're going to deprecate that old behaviour.